✨Update APIVersion for bootstrap references for Machine* resources #1852

neeleshkorade opened the pull request:
**What this PR does / why we need it**:

- Updates `APIVersion` for references to `KubeadmConfig` and `KubeadmConfigTemplate` for Machine* resources. 

Note- Skimming through the code, it appeared to me that it comprised of only two resources- `Machine` and `MachineDeployment`. Please let me know if I have missed out any other resources that need to get this change.

**Which issue(s) this PR fixes**:
Fixes #1852

ncdc commented:
/ok-to-test

Thanks for the PR!

Also need to do this for MachineSets.

ncdc commented on api/v1alpha2/conversion.go:
```diff
@@ -155,6 +155,10 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
 func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *v1alpha3.MachineSpec, s apiconversion.Scope) error {
+	if out.Bootstrap.ConfigRef.Kind == "KubeadmConfig" || out.Bootstrap.ConfigRef.Kind == "KubeadmConfigTemplate" {
```
I would also confirm the .Group matches `bootstrap.kubeadm.api.v1alpha2.GroupVersion.Group` (you can alias the package import to `bootstrapv1`).

neeleshkorade replied:
@ncdc for matching Group, this is what I have done-

ncdc replied:
I meant this:

```go
inGV, err := schema.ParseGroupVersion(in.Boostrap.ConfigRef.APIVersion)
if err != nil {
  return err
}
if inGV.Group == bootstrapv1.GroupVersion.Group && (in.Bootstrap.ConfigRef.Kind == "KubeadmConfig" || in.Bootstrap.ConfigRef.Kind == "KubeadmConfigTemplate") {
  out.Bootstrap.ConfigRef.APIVersion = bootstrapv1.GroupVersion.String()
}
// ...
```

neeleshkorade replied:
Get it. However, `bootstrapv1.GroupVersion.String()` would return `v1alpha2` for the `APIVersion` and not `v1alpha3`, isn't it?

ncdc replied:
It's fine to import the v1alpha3 package as `bootstrapv1` - both v1alpha2 and v1alpha3 have the same API group.

neeleshkorade replied:
Thanks. Updated the PR with the changes.

ncdc commented on api/v1alpha2/conversion.go:
```diff
@@ -164,6 +168,14 @@ func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *
 }
 
+func Convert_v1alpha2_MachineDeploymentSpec_To_v1alpha3_MachineDeploymentSpec(in *MachineDeploymentSpec, out *v1alpha3.MachineDeploymentSpec, s apiconversion.Scope) error {
+	if out.Template.Spec.Bootstrap.ConfigRef.Kind == "KubeadmConfig" || out.Template.Spec.Bootstrap.ConfigRef.Kind == "KubeadmConfigTemplate" {
```
Same comment re group match

neeleshkorade commented:
> /ok-to-test
> 
> Thanks for the PR!
> 
> Also need to do this for MachineSets.



> /ok-to-test
> 
> Thanks for the PR!
> 
> Also need to do this for MachineSets.

Added changes to include `MachineSets`

neeleshkorade commented:
Looking into the test failures.

neeleshkorade commented:
@ncdc also, one of the test cases is failing due to `lint` error-
```
INFO [runner] linters took 13.243718748s with stages: goanalysis_metalinter: 11.582670405s, unused: 1.602888152s
api/v1alpha2/conversion.go:158:88: string `KubeadmConfigTemplate` has 3 occurrences, make it a constant (goconst)
	if out.Bootstrap.ConfigRef.Kind == "KubeadmConfig" || out.Bootstrap.ConfigRef.Kind == "KubeadmConfigTemplate" {
```

Trying to understand what's the best way to fix this. I don't particularly like declaring `struct` names as constants. But I didn't find a pattern elsewhere in the code either. Is there a recommended practice here?

ncdc commented:
You can put

```go
const (
  kubeadmConfigKind = "KubeadmConfig"
  kubeadmConfigTemplateKind = "KubeadmConfigTemplate"
)

in this file if you'd like

ncdc commented:
Thanks! Would you mind adding unit tests that test the changes are correct when converting from alpha2 to alpha3?

@detiber the way this change is implemented, it doesn't guarantee that the resources stored in etcd are modified. It just handles the on the fly conversion if the storage version is alpha2. Although because we do use the deferred PatchHelper for all our resources, they'd all get processed and patched the first time a v1alpha3 pod ran... When the MachineDeployments and MachineSets have their KubeadmConfig refs changed, is that going to trigger a rolling update?

neeleshkorade commented:
@ncdc added unit tests for the three new conversion functions. Please note-

- It does not cover the case where incoming v2 object has `APIVersion` set without `Group` (that is, only `version` is present). My guess is this is not a realistic case (and that is how the new conversion functions have been written at the moment) but let me know if I am mistaken.

detiber commented on api/v1alpha2/conversion.go:
```diff
@@ -155,15 +162,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+	}
+
+	if inGV.Group == bootstrapv1a3.GroupVersion.Group && (out.Template.Spec.Bootstrap.ConfigRef.Kind == kubeadmConfigKind || out.Template.Spec.Bootstrap.ConfigRef.Kind == kubeadmConfigTemplateKind) {
+		out.Template.Spec.Bootstrap.ConfigRef.APIVersion = in.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind().Group + "/" + "v1alpha3"
```
Shouldn't this use `bootstrapv1a3.GroupVersion.String()` like line 175 above?

detiber commented on api/v1alpha2/conversion.go:
```diff
@@ -155,15 +162,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+	}
+
+	if inGV.Group == bootstrapv1a3.GroupVersion.Group && (out.Template.Spec.Bootstrap.ConfigRef.Kind == kubeadmConfigKind || out.Template.Spec.Bootstrap.ConfigRef.Kind == kubeadmConfigTemplateKind) {
+		out.Template.Spec.Bootstrap.ConfigRef.APIVersion = in.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind().Group + "/" + "v1alpha3"
```
Shouldn't this use bootstrapv1a3.GroupVersion.String() like line 175 above?

ncdc commented:
@detiber did you see my question above?

detiber commented:
> @detiber the way this change is implemented, it doesn't guarantee that the resources stored in etcd are modified. It just handles the on the fly conversion if the storage version is alpha2. Although because we do use the deferred PatchHelper for all our resources, they'd all get processed and patched the first time a v1alpha3 pod ran... When the MachineDeployments and MachineSets have their KubeadmConfig refs changed, is that going to trigger a rolling update?

It *should*, however I do wonder if we should also suggest the use of the Storage Version Migrator utility as mentioned here: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version

ncdc commented:
I'm mostly concerned about avoiding unnecessary machine deletions & recreations - will this change result in machines getting deleted and created?

ncdc commented:
I guess that largely depends on the code in the MD/MS controllers and what they consider to be substantive differences.

detiber commented:
> I'm mostly concerned about avoiding unnecessary machine deletions & recreations - will this change result in machines getting deleted and created?

> I guess that largely depends on the code in the MD/MS controllers and what they consider to be substantive differences.

For a MachineSet it wouldn't have any effect. Since it doesn't take into account the configuration when applying changes.

For a MachineDeployment, though I do think it would trigger a rolling deployment based on the way we are currently computing the hash: https://github.com/kubernetes-sigs/cluster-api/blob/fd807a3d843b722a3b4feb3c9e3bcd752896daed/controllers/mdutil/util.go#L674-L693

This is definitely not ideal, and I wonder if we need to change the way we compute the hash to avoid this...  That said, we'd probably want to figure out a way to do that without triggering a new deployment...

neeleshkorade commented:
> A couple of nits around how the API group replacement is being done for MachineSets and MachineDeployments, otherwise lgtm

Thanks for catching it @detiber. I updated the PR.

ncdc commented:
I'm not coming up with any quick ideas, unfortunately

I do not want us to add any logic to the MachineDeployment reconciler that special-cases KubeadmConfigs or KubeadmConfigTemplates.

We potentially could consider excluding the version portion of the `apiVersion` field from the hash, but I'm not sure if that would have unintended consequences (e.g. if a provider does want to roll out a new MachineSet if a reference object's api version changes).

@vincepri wdyt?

vincepri commented:
> We potentially could consider excluding the version portion of the apiVersion field from the hash

This seems a good solution for future iterations, if we do this today, the resulting hash would still be different which would result in a rolling update.

I initially thought we could change the hash during conversion to the new value, and maintain backward compatibility, but I think that would also have the same effect given that MachineSets match on a label that contains the hash. The hash is also used in a MS name to provide idempotency.

neeleshkorade commented:
I haven’t looked that much into the code so I may be totally off but what I gather from the thread is that currently whether to do rolling update or not is purely based on the hash of the objects. Would enhancing the decision making to include cases where rolling update may not be desired even if the object hash changes be an acceptable solution? There might also be a case where even if the hash changes for a reason that entails rolling update, the operator/administrator may want to defer it (something on the lines of the deferred PatchHelper that @ncdc mentioned in one of the comments above).

ncdc commented:
I chatted briefly with @liggitt about this and it's been an open, unresolved issue for at least four years 😦. I think I would prefer to avoid this approach.

Maybe we could consider adding logic to our controllers to pause reconciling, update the object references, and then resume reconciling, ideally with no rolling updates. This might not be possible if the MachineSet name is based on the hash of the template, but it's worth exploring.

neeleshkorade commented:
> I chatted briefly with @liggitt about this and it's been an open, unresolved issue for at least four years 😦. I think I would prefer to avoid this approach.
> 
> Maybe we could consider adding logic to our controllers to pause reconciling, update the object references, and then resume reconciling, ideally with no rolling updates. This might not be possible if the MachineSet name is based on the hash of the template, but it's worth exploring.

@ncdc does this mean this PR gets closed and a different approach is to be taken that takes care of updating versions while preventing any redundant rolling updates from happening?

ncdc commented:
@neeleshkorade sorry for the delay. I _think_ we may have a path forward. We started a discussion in Slack (https://kubernetes.slack.com/archives/C8TSNPY4T/p1580231398178200) and I'll update here once we have a chance to discuss in more detail.

ncdc commented:
Update: we're going to do #2183, and then this PR can merge.

neeleshkorade commented:
Thanks for the update @ncdc

vincepri commented:
@neeleshkorade Are you able to rebase on top of master? We should be able to get these changes in once the rebase has been done.

neeleshkorade commented:
> @neeleshkorade Are you able to rebase on top of master? We should be able to get these changes in once the rebase has been done.

@vincepri I just rebased and updated the PR. Waiting for tests to go through. One thing though- the `convertFrom` part in the newly added tests `TestConvertMachine`, `TestConvertMachineSet` and `TestConvertMachineDeployment` failed. The reason is that the v3 to v2 conversion functions (for `Machine`, `MachineSet` and `MachineDeployment`) do not seem to copy over `spec` field. This was leading to "nil pointer dereference" errors at places like- https://github.com/kubernetes-sigs/cluster-api/blob/68f64459119fb45922f3a3c629d774b4eb985abf/api/v1alpha2/conversion.go#L293

For now, I have hacked the failing tests a bit like here- https://github.com/kubernetes-sigs/cluster-api/pull/1938/files#diff-c8f1f9d0a7bf481d8e912ebc493d3c9fR165-R169

I am trying to figure if my understanding is correct and if so where the changes should be made.

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+		return err
+	}
+
+	if err = autoConvert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec(in, out, s); err != nil {
```
```suggestion
	if err := autoConvert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec(in, out, s); err != nil {
```

vincepri replied:
This should be the first thing called

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
+func Convert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec(in *MachineSetSpec, out *v1alpha3.MachineSetSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Template.Spec.Bootstrap.ConfigRef.APIVersion)
```
This should be guarded against `in.Template.Spec.Bootstrap != nil`, move this at the bottom

neeleshkorade replied:
If `in.Template.Spec.Bootstrap == nil`, is it a normal case or an error condition? I earlier thought of doing this but somehow thought except for objects created in tests, the input `in` objects should always have non-nil `in.Template.Spec.Bootstrap`. But may be I am wrong.

vincepri replied:
If the Bootstrap is nil that's a normal condition, and no further action is necessary

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
 func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *v1alpha3.MachineSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Bootstrap.ConfigRef.APIVersion)
```
Move this down, and guard against `in.Bootstrap.ConfigRef` being != nil

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+		return err
+	}
+
 	if err := autoConvert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in, out, s); err != nil {
```
This should be the first call

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+		return err
+	}
+
+	if err = autoConvert_v1alpha2_MachineDeploymentSpec_To_v1alpha3_MachineDeploymentSpec(in, out, s); err != nil {
```
```suggestion
	if err := autoConvert_v1alpha2_MachineDeploymentSpec_To_v1alpha3_MachineDeploymentSpec(in, out, s); err != nil {
```

This should be the first call

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
+}
+
+func Convert_v1alpha2_MachineDeploymentSpec_To_v1alpha3_MachineDeploymentSpec(in *MachineDeploymentSpec, out *v1alpha3.MachineDeploymentSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Template.Spec.Bootstrap.ConfigRef.APIVersion)
```
Move this down, and guard against in.Bootstrap.ConfigRef being != nil

vincepri commented on api/v1alpha2/conversion.go:
```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 
+	if inGV.Group == bootstrapv1a3.GroupVersion.Group && (out.Bootstrap.ConfigRef.Kind == kubeadmConfigKind || out.Bootstrap.ConfigRef.Kind == kubeadmConfigTemplateKind) {
+		out.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
+	}
```
> NB: This comment applies to all conversions.

This and the check above can be simplified:

```golang
if in.Bootstrap.ConfigRef != nil {
		gvk := in.Bootstrap.ConfigRef.GetObjectKind().GroupVersionKind()
		if gvk.Group == bootstrapv1a3.GroupVersion.Group && gvk.Kind == kubeadmConfigTemplateKind {
			out.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
		}
	}
```

The template is the only reference we should change for MachineSet or MachineDeployment, while for Machines we only need to check the `KubeadmConfig` kind.

neeleshkorade replied:
> The template is the only reference we should change for MachineSet or MachineDeployment, while for Machines we only need to check the `KubeadmConfig` kind.

Based on this, now
-  for `MachineSpec`, checking only for `kubeadmConfigKind`
- for `MachineSetSpec` and `MachineDeploymentSpec`, checking only for `kubeadmConfigTemplateKind`

Also, based on my earlier understanding, I added tests for the conditions when a `MachineSet` object has `kubeadmConfigTemplate` kind reference and `MachineSetSpec` and `MachineDeploymentSpec` objects have `kubeadmConfig` kind references. For example- https://github.com/kubernetes-sigs/cluster-api/pull/1938/files#diff-c8f1f9d0a7bf481d8e912ebc493d3c9fR337

vincepri commented:
@neeleshkorade Once the fix suggested above is implemented, the tests shouldn't fail anymore and you can remove the hack. Note that the destination is meant to be an empty object, shouldn't have any data prior to conversion.

neeleshkorade commented:
> @neeleshkorade Once the fix suggested above is implemented, the tests shouldn't fail anymore and you can remove the hack. Note that the destination is meant to be an empty object, shouldn't have any data prior to conversion.

@vincepri have made the changes you suggested. The tests are passing now in my local run. Waiting for it to succeed in CI.

vincepri commented on api/v1alpha2/conversion_test.go:
```diff
@@ -243,3 +282,150 @@ func TestConvertMachineDeployment(t *testing.T) {
 	})
 }
+
+func TestConvert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(t *testing.T) {
```
Tests should be rewritten using the gomega library as shown above with other tests

neeleshkorade replied:
Made the change to use gomega for assertions. Please let me know if anything else is missing.

ncdc replied:
Please remove all these new test functions. The suggestions I made above cover everything you've added here.

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -24,6 +24,21 @@ import (
+	}
+	verV1alpha2 = bootstrapv1a2.GroupVersion.String()
+	verV1alpha3 = bootstrapv1a3.GroupVersion.String()
 )
```
We can't have the object references as package level variables. If you use them with the objects being tested, they get modified by the conversion code, and the tests "pass" but they're not testing the correct values.

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -99,12 +114,18 @@ func TestConvertMachine(t *testing.T) {
 
 	t.Run("to hub", func(t *testing.T) {
 		t.Run("should convert the Spec.ClusterName from label", func(t *testing.T) {
+
```
```suggestion
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -99,12 +114,18 @@ func TestConvertMachine(t *testing.T) {
 				},
+				Spec: MachineSpec{
+					Bootstrap: Bootstrap{
+						ConfigRef: kubeAdmObj,
```
```suggestion
						ConfigRef: &corev1.ObjectReference{
							Kind:       kubeadmConfigKind,
							APIVersion: bootstrapv1a2.GroupVersion.String(),
						},
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -99,12 +114,18 @@ func TestConvertMachine(t *testing.T) {
 
 	t.Run("to hub", func(t *testing.T) {
 		t.Run("should convert the Spec.ClusterName from label", func(t *testing.T) {
```
```suggestion
		t.Run("should convert all fields correctly", func(t *testing.T) {
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -99,12 +114,18 @@ func TestConvertMachine(t *testing.T) {
+					Bootstrap: Bootstrap{
+						ConfigRef: kubeAdmObj,
+					},
+				},
```
I can't comment on the correct line, but please add this after the last expectation in this test case:

```go
g.Expect(dst.Spec.Bootstrap.ConfigRef.APIVersion).To(Equal(bootstrapv1a3.GroupVersion.String()))
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -154,6 +175,15 @@ func TestConvertMachineSet(t *testing.T) {
 						MachineClusterLabelName: "test-cluster",
 					},
 				},
+				Spec: MachineSetSpec{
```
Please adjust the test case name to "should convert all fields correctly" above

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -154,6 +175,15 @@ func TestConvertMachineSet(t *testing.T) {
+					Template: MachineTemplateSpec{
+						Spec: MachineSpec{
+							Bootstrap: Bootstrap{
+								ConfigRef: kubeAdmObj,
```
```suggestion
								ConfigRef: &corev1.ObjectReference{
									Kind:       kubeadmConfigTemplateKind,
									APIVersion: bootstrapv1a2.GroupVersion.String(),
								},
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -154,6 +175,15 @@ func TestConvertMachineSet(t *testing.T) {
+							},
+						},
+					},
+				},
```
Please add this expectation:

```go
g.Expect(dst.Spec.Template.Spec.Bootstrap.ConfigRef.APIVersion).To(Equal(bootstrapv1a3.GroupVersion.String()))
```

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -203,6 +233,16 @@ func TestConvertMachineDeployment(t *testing.T) {
+					Template: MachineTemplateSpec{
+						Spec: MachineSpec{
+							Bootstrap: Bootstrap{
+								ConfigRef: kubeAdmObj,
```
```suggestion
								ConfigRef: &corev1.ObjectReference{
									Kind:       kubeadmConfigTemplateKind,
									APIVersion: bootstrapv1a2.GroupVersion.String(),
								},
```

neeleshkorade replied:
Done

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -203,6 +233,16 @@ func TestConvertMachineDeployment(t *testing.T) {
 						MachineClusterLabelName: "test-cluster",
 					},
 				},
+				Spec: MachineDeploymentSpec{
```
Please update the test case name as above

ncdc replied:
^^^^^

neeleshkorade replied:
Updated `should convert all fields correctly`

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -203,6 +233,16 @@ func TestConvertMachineDeployment(t *testing.T) {
+							},
+						},
+					},
+				},
```
Please add this expectation below:
```go
g.Expect(dst.Spec.Template.Spec.Bootstrap.ConfigRef.APIVersion).To(Equal(bootstrapv1a3.GroupVersion.String()))
```

neeleshkorade commented:
@ncdc made the changes you suggested. Please take a look

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -105,11 +107,20 @@ func TestConvertMachine(t *testing.T) {
 						MachineClusterLabelName: "test-cluster",
 					},
 				},
+				Spec: MachineSpec{
```
Can you adjust the name of this test case too please?

neeleshkorade replied:
Updated to `should convert all fields correctly`

neeleshkorade replied:
@ncdc unable to comment on the line so pasting link here. Should this test case name be changed too? https://github.com/kubernetes-sigs/cluster-api/pull/1938/files#diff-c8f1f9d0a7bf481d8e912ebc493d3c9fR35

ncdc replied:
That's linking to a line that's not part of a test case, but we're good enough for now.

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -192,7 +216,7 @@ func TestConvertMachineSet(t *testing.T) {
 }
 
-func TestConvertMachineDeployment(t *testing.T) {
+func TestConvertMachineDeploymentSpec(t *testing.T) {
```
Please undo this change

neeleshkorade replied:
Done

ncdc commented on api/v1alpha2/conversion_test.go:
```diff
@@ -242,4 +279,4 @@ func TestConvertMachineDeployment(t *testing.T) {
 		})
 	})
-}
+}
```
Please don't remove the newline at the end of the file. Thanks.

neeleshkorade replied:
Done

ncdc commented on hack/tools/go.mod:
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/blang/semver v3.5.1+incompatible
 	github.com/golangci/golangci-lint v1.21.0
 	github.com/jteeuwen/go-bindata v3.0.7+incompatible
+	golang.org/x/exp v0.0.0-20190312203227-4b39c73a6495 // indirect
```
Do you know where this is coming from? cc @vincepri

ncdc replied:
Looks like you need to run `make modules`

vincepri replied:
I'm not totally sure, we shouldn't be using anything from the experimental go package afaik, unless we mistakenly updated some deps

neeleshkorade replied:
Ran `make verify-modules` and it removed the redundant module. Pushed the change, waiting for the build to go through.

ncdc commented:
@neeleshkorade there are still outstanding comments - will you be able to address them?

ncdc commented:
Please squash and rebase on top of master - thanks!

neeleshkorade commented:
@ncdc running into this issue- `make verify-gen` fails if `golang.org/x/exp v0.0.0-20190312203227-4b39c73a6495 // indirect` `require` dependency is missing in `hack/tools/go.mod`, while `make verify-modules` fails if the dependency is there. Any idea how to fix this?

neeleshkorade commented:
/retest

detiber commented on api/v1alpha2/conversion.go:
```diff
@@ -267,11 +273,48 @@ func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *
+		if gvk.Group == bootstrapv1a3.GroupVersion.Group && gvk.Kind == kubeadmConfigTemplateKind {
+			out.Template.Spec.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
+		}
+	}
```
I'm actually wondering if this is strictly needed here, the `autoConvert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec` method will cause `Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec` to be called.

vincepri replied:
It should be easy to check by removing this line, the test cases should still pass

neeleshkorade replied:
This works for `Convert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec` but not for `Convert_v1alpha2_MachineDeploymentSpec_To_v1alpha3_MachineDeploymentSpec` (the test `TestConvertMachineDeployment` fails).

Looks like the reason is from [here](https://github.com/kubernetes-sigs/cluster-api/blob/ddefb05f0597372c53b1a051b7b544e4c3c44408/api/v1alpha2/conversion.go#L278). We updated the `if` condition to exclude `kubeadmConfigTemplateKind` so now it fails for `MachineDeploymentSpec` objects. Updating the `if` condition makes the test pass.

detiber commented on api/v1alpha2/conversion.go:
```diff
@@ -267,11 +273,48 @@ func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *
+		if gvk.Group == bootstrapv1a3.GroupVersion.Group && gvk.Kind == kubeadmConfigTemplateKind {
+			out.Template.Spec.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
+		}
+	}
```
I'm actually wondering if this is strictly needed here, the autoConvert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec method will cause Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec to be called.

vincepri commented:
@neeleshkorade Thinking more about this problem, it seems that this approach might not be what we want. 

Going forward, we'll have support to understand contracts with:
- A generic solution that understands API contracts for any external reference https://github.com/kubernetes-sigs/cluster-api/pull/2267
- Moving away from ObjectReference (and possibly drop the version field altogether) with https://github.com/kubernetes-sigs/cluster-api/issues/2318

neeleshkorade commented:
> @neeleshkorade Thinking more about this problem, it seems that this approach might not be what we want.
> 
> Going forward, we'll have support to understand contracts with:
> 
> * A generic solution that understands API contracts for any external reference #2267
> * Moving away from ObjectReference (and possibly drop the version field altogether) with #2318

Thanks @vincepri I went through the referenced issues and get what you are saying. Let me know if you would like me to close this PR.

vincepri commented:
Thanks again for understanding!

/close
